server/metadata/queue.go

```diff
@@ -0,0 +1,241 @@
+func (q *QueueSubspace) Enqueue(ctx context.Context, tx transaction.Tx, item *QueueItem, delay time.Duration) error {
+	if len(item.Id) == 0 {
+		return errors.Internal("cannot enqueue without an id")
+	}
+	item.VestingTime = time.Now().Add(delay).UnixMilli()
+	key := q.getKey(item.VestingTime, item.Priority, item.Id)
+	tableData, err := q.encodeItem(item)
+	if err != nil {
+		return err
+	}
+	return tx.Replace(ctx, key, tableData, false)
+}
+
+func (q *QueueSubspace) Peak(ctx context.Context, tx transaction.Tx, max int) ([]QueueItem, error) {
+	var items []QueueItem
+	count := 0
+	currentTime := time.Now().UnixMilli()
+	endKey := q.getKey(currentTime, maxPriority, maxId)
+	iter, err := tx.ReadRange(ctx, nil, endKey, false)
+	if err != nil {
+		return items, err
+	}
+	var v kv.KeyValue
+	for iter.Next(&v) && count < max {
+		count += 1
+		item, err := q.decodeItem(v.Data)
+		if err != nil {
+			return nil, err
+		}
+		items = append(items, *item)
+	}
+
+	return items, iter.Err()
+}
+
+func (q *QueueSubspace) Find(ctx context.Context, tx transaction.Tx, item *QueueItem) (*QueueItem, error) {
+	startKey := q.getKey(item.VestingTime, 0, item.Id)
+	endKey := q.getKey(item.VestingTime, maxPriority, item.Id)
+
+	iter, err := tx.ReadRange(ctx, startKey, endKey, false)
+	if err != nil {
+		return nil, err
+	}
+	var v kv.KeyValue
+	for iter.Next(&v) {
+		found, err := q.decodeItem(v.Data)
+		if err != nil {
+			return nil, err
+		}
+
+		if item.Id == found.Id {
+			return item, nil
+		}
+	}
+
+	if iter.Err() != nil {
+		return nil, iter.Err()
+	}
+	return nil, errors.Internal("QueueItem \"%s\" was not found", item.Id)
+}
+
+func (q *QueueSubspace) Dequeue(ctx context.Context, tx transaction.Tx, item *QueueItem) error {
+	itemKey := q.getKey(item.VestingTime, item.Priority, item.Id)
+	return tx.Delete(ctx, itemKey)
+}
+
+func (q *QueueSubspace) ObtainLease(ctx context.Context, tx transaction.Tx, item *QueueItem, leaseTime time.Duration) (*QueueItem, error) {
+	item, err := q.Find(ctx, tx, item)
+	if err != nil {
+		return nil, err
+	}
+	if item == nil {
+		return nil, errors.Internal("Queue Item was not found")
+	}
+
+	if err = q.Dequeue(ctx, tx, item); err != nil {
+		return nil, err
+	}
+
+	item.LeaseId = uuid.NewUUIDAsString()
+	if err = q.Enqueue(ctx, tx, item, leaseTime); ulog.E(err) {
+		return nil, err
+	}
+
+	return item, nil
+}
+
+func (q *QueueSubspace) Complete(ctx context.Context, tx transaction.Tx, item *QueueItem) error {
+	found, err := q.Find(ctx, tx, item)
+	if err != nil {
+		return err
+	}
+
+	if found.LeaseId != item.LeaseId {
+		return errors.Internal("another worker has claimed this item")
+	}
+
+	return q.Dequeue(ctx, tx, item)
+}
+
+func (q *QueueSubspace) RenewLease(ctx context.Context, tx transaction.Tx, item *QueueItem, leaseTime time.Duration) error {
+	found, err := q.Find(ctx, tx, item)
+	if err != nil {
+		return err
+	}
+
+	if found.LeaseId != item.LeaseId {
+		return errors.Internal("another worker has claimed this item")
```

Comment:
So these errors doesn't need to be HTTP errors, we can have some errors defined for this class which can be used in this case. For example, this error is just throw away error i.e. if item is picked by some other error then ignore the item and continue. But in some cases the errors will be critical and we may need additional actions.